credentials/credentials.go

```diff
@@ -161,17 +147,18 @@ func (c *tlsCreds) ClientHandshake(addr string, rawConn net.Conn, timeout time.D
 		cfg.ServerName = addr[:colonPos]
 	}
 	conn := tls.Client(rawConn, cfg)
-	if timeout == 0 {
-		err = conn.Handshake()
-	} else {
-		go func() {
-			errChannel <- conn.Handshake()
-		}()
-		err = <-errChannel
-	}
-	if err != nil {
-		rawConn.Close()
-		return nil, nil, err
+	errChannel := make(chan error, 1)
+	go func() {
+		errChannel <- conn.Handshake()
+	}()
+	select {
+	case err := <-errChannel:
+		if err != nil {
+			rawConn.Close()
+			return nil, nil, err
+		}
+	case <-ctx.Done():
```

Comment:
Sure. I was going for the publicly-visible description, since ClientHandshake is the only public interface changed. What would you like it to say?